### PR TITLE
[Core] order-by support

### LINKED ISSUE #241: Ordering search results
Hey everybody, amazing library!

I'm looking to build something similar to the Jira extended search input field for issues:
bildschirmfoto 2018-10-19 um 09 30 48

It looks like rollerworks/search is already half way there! I can't find a way how to order the search results though. Is it planned to add anything like that? That would be amazing.

### PR DESCRIPTION
**Pull request by dkarlovi**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #241 
| License       | MIT
| Doc PR        | N/A

Closes #241.

### CODE REVIEW

File: lib/Core/Field/GenericResolvedFieldType.php [hunk 1]
```diff
@@ -145,6 +145,10 @@ public function getOptionsResolver(): OptionsResolver
      */
     protected function newField($name, array $options): FieldConfig
     {
+        if ('@' === $name[0]) {
```
Comment (dkarlovi) [inline lib/Core/Field/GenericResolvedFieldType.php]: @sstok does this seem correct to you? I'm kind of lost in the complexity, TBH. :see_no_evil:

File: lib/Core/Field/OrderFieldType.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return null;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        // TODO: Implement configureOptions() method.
```
Comment (dkarlovi) [inline lib/Core/Field/OrderFieldType.php]: Validators, etc.

File: lib/Core/Input/ConditionStructureBuilder.php [hunk 3]
```diff
@@ -165,7 +165,9 @@ public function field(string $name, string $path): void
         $this->fieldConfig = $this->fieldSet->get($name);
         $this->valuesBag = new ValuesBag();
 
-        $this->valuesGroupLevels[$this->nestingLevel]->addField($name, $this->valuesBag);
+        $normalizedName = '@' !== $name[0] ? $name : substr($name, 1);
```
Comment (dkarlovi) [inline lib/Core/Input/ConditionStructureBuilder.php]: Not happy about this.

File: lib/Core/Input/JsonInput.php [hunk 4]
```diff
@@ -208,6 +212,26 @@ private function processFields(array $values)
         }
     }
 
+    private function processOrder(SearchCondition $condition, $array)
+    {
+        $order = $array['order'] ?? [];
```
Comment (dkarlovi) [inline lib/Core/Input/JsonInput.php]: Switch to the same approach as in in StringInput, separate conditions builder for order?

File: lib/Core/Tests/Input/JsonInputTest.php [hunk 5]
```diff
@@ -66,6 +66,9 @@ public function provideSingleValuePairTests()
                                 'excluded-simple-values' => ['value3'],
                             ],
                         ],
+                        'order' => [
+                            '@date' => 'ASC',
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/JsonInputTest.php]: Do we need to enforce the fields are always prefixed with `@` here, do we do it ourselves in the JsonInput?

File: lib/Core/Field/GenericResolvedFieldType.php (hunk 1 above)
Comment (sstok) [reply lib/Core/Field/GenericResolvedFieldType.php]: What do you mean exactly? The CS?

File: lib/Core/Tests/Input/JsonInputTest.php (hunk 5 above)
Comment (sstok) [reply lib/Core/Tests/Input/JsonInputTest.php]: I think we can handle this automatically, no need to expose internal details here.

File: lib/Core/Field/GenericResolvedFieldType.php (hunk 1 above)
Comment (dkarlovi) [reply lib/Core/Field/GenericResolvedFieldType.php]: No, the way I've done it here. It looks a bit smelly, sniffing the name everywhere.

**Comment by dkarlovi** [comment]
@sstok @toflar ready for review

**Comment by dkarlovi** [comment]
Just figured out another use case, working on it.

File: lib/Core/Tests/Input/InputProcessorTestCase.php [hunk 6]
```diff
@@ -51,6 +54,10 @@ protected function getFieldSet(bool $build = true)
         $fieldSet->add('name', TextType::class);
         $fieldSet->add('lastname', TextType::class);
         $fieldSet->add('date', DateType::class, ['pattern' => 'MM-dd-yyyy']);
+        if ($order) {
+            $fieldSet->add('@date', OrderFieldType::class, ['case' => OrderTransformer::CASE_LOWERCASE,  'alias' => ['up' => 'ASC', 'down' => 'DESC'], 'default' => 'down']);
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/InputProcessorTestCase.php]: @toflar @sstok this is an example usage.

File: lib/Core/Tests/Input/JsonInputTest.php [hunk 7]
```diff
@@ -57,6 +57,22 @@ public function provideEmptyInputTests()
...
+                                'simple-values' => ['value', 'value2', '٤٤٤٦٥٤٦٠٠', '30', '30L'],
+                                'excluded-simple-values' => ['value3'],
+                            ],
+                        ],
+                        'order' => [
+                            'date' => 'uP',
+                        ],
+                    ]
+                ),
+                ['@date' => 'ASC'],
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/JsonInputTest.php]: Normalized from user input

File: lib/Core/Tests/Input/JsonInputTest.php [hunk 8]
```diff
@@ -68,6 +84,7 @@ public function provideSingleValuePairTests()
                         ],
                     ]
                 ),
+                ['@date' => 'DESC', '@id' => 'ASC'],
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/JsonInputTest.php]: Default sorts, only if the user hasn't specified anything.

**Comment by dkarlovi** [comment]
To clarify the functionality:

- we're using the same prefix `@` for marking the order-by field as we're using in string input (but only by convention)
- we're using this to be able to specify a different mapping for searching a field and sorting on a field (for example, in Elasticsearch it would be `title: title, @title: title.keyword`)
- you can specify a normalization for the direction like `LOWERCASE`, `UPPERCASE` (default)
- you can specify a custom map of view value => model value, default is `ASC => ASC, DESC => DESC`, this allows you to specify additional values and their "meanings", change case if your backend requires it, etc.

TLDR

```
case = LOWERCASE, map = [up => ASC, down => DESC], query = @date: uP, condition = 
order[@date: ASC]
```

**Comment by Toflar** [comment]
Looks awesome to me. A bit hard for me to review because I'm quite new to this library but at first sight I can't see any conceptual issues here. The only thing that comes to my mind is the question whether allowed field name characters have been validated so far or if prefixing by an `@` introduces a potential BC break. Not that a think there have been many out there using an `@` prefix but it would maybe be a good idea to restrict the field names if not done already (just double checking).

Also, it seems to me like the current implementation was not locale aware at all. With the introduction of aliases (e.g. like in your example) it kind of becomes localized. `@field up` is English and probably devs will start adding aliases in different languages which seems strange to me for a format that was independent from any language so far. Just a thought though 😄

**Comment by dkarlovi** [comment]
@Toflar thanks for your feedback.

> only thing that comes to my mind is the question whether allowed field name characters have been validated so far

It has, @ has not been allowed by `StringLexer`, I've allowed for it here.

> it would maybe be a good idea to restrict the field names if not done already

You mean, you can only bind the `@` prefix with a `OrderFieldType`? That could be done. @sstok WDYT?

> With the introduction of aliases (e.g. like in your example) it kind of becomes localized. @field up is English and probably devs will start adding aliases in different languages

The introduction of alliases allows for several things:

1. allow **whitelisting** directions, for example, specifying only `['ASC => 'ASC'']` means you're disallowing `DESC` completely.
2. allow tweaking the **model**, for example if in a backend you need to say `DESCENDING` instead of just `DESC`, this would allow you to specify the final version you need
3. allow tweaking the **view** (which you note), this is from my POV the only questionable point, do we want this?

Let's wait and see what @sstok says.

**Comment by dkarlovi** [comment]
BTW note one thing: order's... order :see_no_evil: is **significant**.

This means that `@foo: DESC; @bar: ASC` and `@bar: ASC; @foo: DESC` are **not interchangeable**. The order clause will be created in the same order the user has specified it. We could also opt to always create in the order the fields were added to the fieldset (so, above would be interchangeable, but you lose a bit of flexibility).

This is also up for discussion.

**Comment by dkarlovi** [comment]
Agree, we should use user input, but it's something to note so it's not surprising.

**Comment by dkarlovi** [comment]
@sstok it would be great if you could do a review by Thursday, that's when I'll continue working on this and #256.

File: lib/Core/Input/StringInput.php [hunk 9]
```diff
@@ -275,7 +302,14 @@ private function fieldGroup(): void
 
     private function fieldValues(string $name): void
     {
-        $this->structureBuilder->field($name, '[%s]');
+        $isOrder = '@' === $name[0];
+
+        $this->hasOrder = $this->hasOrder || $isOrder;
+
+        /** @var ConditionStructureBuilder $structureBuilder */
+        $structureBuilder = $isOrder ? $this->orderStructureBuilder : $this->structureBuilder;
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply lib/Core/Input/StringInput.php]: All true, thanks! :+1:

File: lib/Core/Input/StringInput.php [hunk 10]
```diff
@@ -275,7 +302,14 @@ private function fieldGroup(): void
 
     private function fieldValues(string $name): void
     {
-        $this->structureBuilder->field($name, '[%s]');
+        $isOrder = '@' === $name[0];
+
+        $this->hasOrder = $this->hasOrder || $isOrder;
```
[the remark this answers is not in the document]
Comment (dkarlovi) [reply lib/Core/Input/StringInput.php]: Excellent point! :+1:

**Comment by dkarlovi** [comment]
@sstok had to squash since the interface change was the first commit. I'll add the separate structure builder ASAP.

File: lib/Core/Input/StringInput.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (dkarlovi) [reply lib/Core/Input/StringInput.php]: @sstok added `OrderStructureBuilder`.

File: lib/Core/Tests/Input/StringQueryInputTest.php [hunk 11]
```diff
@@ -319,8 +319,10 @@ public function provideEmptyInputTests()
     public function provideSingleValuePairTests()
     {
         return [
-            ['name: value, value2, ٤٤٤٦٥٤٦٠٠, 30, 30L, !value3;'],
-            ['name: value, value2, ٤٤٤٦٥٤٦٠٠, 30, 30L, !value3'],
+            ['name: value, value2, ٤٤٤٦٥٤٦٠٠, 30, 30L, !value3; @date: uP', ['@date' => 'ASC']],
+            ['name: value, value2, ٤٤٤٦٥٤٦٠٠, 30, 30L, !value3; @date: Up', ['@date' => 'ASC']],
+            ['name: value, value2, ٤٤٤٦٥٤٦٠٠, 30, 30L, !value3;', ['@date' => 'DESC', '@id' => 'ASC']],
+            ['', ['@date' => 'DESC', '@id' => 'ASC']],
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/StringQueryInputTest.php]: Default sort options still get applied, even if the search & filter part is empty.

File: lib/Core/Tests/Input/JsonInputTest.php [hunk 12]
```diff
@@ -66,8 +66,31 @@ public function provideSingleValuePairTests()
...
+                        ],
+                    ]
+                ),
+                ['@date' => 'DESC', '@id' => 'ASC'],
+            ],
+            [
+                json_encode(
+                    []
+                ),
+                ['@date' => 'DESC', '@id' => 'ASC'],
```
Comment (dkarlovi) [inline lib/Core/Tests/Input/JsonInputTest.php]: Default sort options still get applied, even if the search & filter part is empty.

**Comment by sstok** [comment]
Thank you @dkarlovi
